### ISSUE TITLE
dnsdist:StatNodeStats does not show consistent stats for SuffixMatchRule visitor function

### ISSUE THREAD

**Issue by jacobbunk** [PowerDNS/pdns#13017]
- Program: dnsdist 
 - Issue type: Bug report

### Short description
Once a SuffixMatchRule has triggered a dynamic block the StatNodeStats objects that get passed to the visitor function are no longer updated correctly, causing the dynamic block to not get renewed as expected.

### Environment

 - Operating system: Ubuntu 22.04
 - Software version: dnsdist 1.8
 - Software source: PowerDNS apt repo.

### Steps to reproduce
Create a config with something like:

```
local dbr = dynBlockRulesGroup()
dbr:setSuffixMatchRule(10, 'Exceeded query limit', 30, DNSAction.Drop, visitor)

function visitor(n, s, c)
  if c.queries > 5
     return true
  end
  return false
end
```
Send it a bunch of queries for a specific domain to trigger a dynamic block. Once the block is in place `c` is no longer updated and the block will expire after 30s even if the high query rate persists.

### Expected behaviour
The incoming queries that does get blocked are still reflected in the output of topQueries(), so I would expect it to also be reflected in the StatNodeStats objects passed to the visitor function.

### Actual behaviour
The StatNodeStats objects appears to no longer reflect the incoming queries that are being blocked by the dynamic block.

**Comment by rgacogne** [comment]
The `StatNode` object is built by looking at responses, not queries, because some information is only available from the response (rcode, response size, cache hit or miss, ...), and as soon as a domain is blocked there is by definition no more responses sent for it.
I think we need to document this clearly, and perhaps there is a valid need for a separate visitor function executed on queries, but I don't think we can change the behaviour of this one.

**Comment by jacobbunk** [comment]
If I create a dynamic block based on query rates from individual clients like:
```
dbr:setQueryRate(5, 10, 'Exceeded query rate', 60)
```
The dynamic block will automatically renew if the query rate continues to be high from that client. I want to do something similar for a `SuffixMatchRule()`. I do understand that the `nxdomains`, `hits`, `servfails`, `bytes` etc. will depend on responses, but it's confusing, to me at least, that the `queries` statistic is not updated to be consisted with the output of `topQueries()`.

**Comment by rgacogne** [comment]
Yes, `setQueryRate` indeed looks at queries (and queries only), which is why it keeps seeing queries even if they are blocked.
I do understand why you would want the `SuffixMatchRule` visitor to work that way, but it was really designed to look at responses, not queries, and in most cases scanning the queries ring buffer in addition to the responses ring buffer, which would be needed to get the behaviour you want, would be a quite noticeable waste of CPU cycles for large ring buffers. So I'll think about it, we might be able to make that behaviour configurable, but I don't want to change the default behaviour.
